src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+    ///
+    ///     // Cast nanoseconds to seconds to int.
+    ///     Nanoseconds timestamp{1000000000}; // This equals one second.
+    ///     int time_casted = castTime<int, Seconds>(timestamp);
+    ///     ---> time_casted = 1.
+    template<typename T, typename DCast, typename DIn> 
+    T castTime(const DIn &timestamp);
+    template<typename T, typename D> 
+    T castTime(const D &timestamp);
+
+    /// Sets a new week number for the timestamp.
+    ///
+    /// The resulting time since epoch is calculated using the old time of week and 
+    /// the new week number.
+    ///
+    /// Example usage:
+    ///     // The timestamp will now be different! It will be the result from setting 
+    ///     // the new week number to 4.
+    ///     Seconds timestamp = TimestampState{}.getTimestamp<Seconds>();
+    ///     setWeek(timestamp, 4);
+    template<typename D>
+    void setWeek(D &timestamp, int week);
+
+    /// Sets a new time of week for the timestamp.
+    ///
+    /// The resulting time since epoch is calculated using the old week number and 
+    /// the new time of week.
+    ///
+    /// Example usage:
+    ///     // Setting time of week with the same unit duration as the timestamp.
+    ///     Seconds timestamp = TimestampState{}.getTimestamp<Seconds>();
+    ///     setTimeOfWeek<Seconds>(timestamp, 300);
+    ///     // ---> The timestamp will now be different! It will be the result from 
+    ///     //      setting the new time of week to 300 seconds. 
+    ///
+    ///     // Setting time of week with a different unit duration from the timestamp.
+    ///     Nanoseconds timestamp = TimestampState{}.getTimestamp<Nanoseconds>();
+    ///     setTimeOfWeek<Seconds>(timestamp, 6);
+    ///     // ---> The timestamp will now be the result from setting the new time of 
+    ///     //      week to 6 seconds (6,000,000,000 nanoseconds).
+    template<typename DTimeSet, typename DIn> 
+    void setTimeOfWeek(DIn &timestamp, int time);
+    
+    // TODO: Add manual timestamp update function.
+
+
+    /**************************************************************************************/
+    /* NOTE: The following are definitions for all template declarations above. There are */
+    /*       no new declarations following this statement.                                */
+    /**************************************************************************************/
+
+    template<typename D> 
+    inline D getTimeOfWeek(const D &timestamp)
+    {
+        if (timestamp <= Weeks{1}) {return timestamp;}
+
+        return timestamp - duration_cast<Weeks>(timestamp);
+    }
+
+    template<typename DCompare, typename D1, typename D2> 
+    inline bool timeChanged(const D1 &timestamp1, const D2 &timestamp2)
+    {
+        if (timestamp1 >= timestamp2)
+            return (timestamp1 - timestamp2 >= DCompare{1}) ? true : false;
+
+        return (timestamp2 - timestamp1 >= DCompare{1}) ? true : false;
+    }
```

Comment:
The function is checking for a difference in the time unit like you said in the second comment. Since there is a check for which one is greater, the second duration will be converted to the greater one before it is subtracted. So for example:
```
Minutes greater(2);
Seconds less(60);
// Comparing on Milliseconds evaluates to: Milliseconds(Minutes(2) - Minutes(1)) >= Milliseconds(60,000) --> true.
```

If it has changed, it will always be >= 1. It can't be less than 1 (unless we support fractional values). Here's another example:
```
Seconds greater(121);
Minutes less(2);
// Comparing on Seconds evaluates to: Seconds(Seconds(121) - Seconds(120)) >= Seconds(1) --> true.
```

What do you think about that? Does that help paint a different picture or is the interface too confusing?